lib/edit-site-page-templates.php

```diff
@@ -0,0 +1,47 @@
+<?php
+/**
+ * REST endpoint for exposing theme's templates.
+ *
+ * @package gutenberg
+ */
+
+function gutenberg_edit_site_get_page_templates() {
```

Comment:
I'm not totally sure we need this. Does this existing code help?

https://github.com/WordPress/gutenberg/blob/848b4934242512ab3909e9851842dfac4732512a/lib/templates.php#L179-L201

There is another example here which is used for edit site export:

https://github.com/WordPress/gutenberg/blob/848b4934242512ab3909e9851842dfac4732512a/lib/edit-site-export.php#L21-L38

In each of those cases, I think we go through each of the available template types, and try to resolve templates for them, rather than looking through the theme files. IIRC, that code also handles creating autodrafts of templates which don't exist.

I might be wrong though, it's a confusing part of the codebase :)

Reply:
My code is putting together an array of file paths of "physical template files". So these aren't stored as a `wp_template` in the database so we can't request it with REST API.

`filter_rest_wp_template_query` to utilise this, we need an actual template type that we can query. But we don't know those. That's why we need this little endpoint I created.

`auto-drafts` are created once the template is selected for editing.

Reply:
I'm trying to wrap my head around all of this, so apologize in advance if I'm missing the point, or saying something really dumb.

As far as I understand, these autodrafts of templates and template parts are created within the [`rest_{$this->post_type}_query`](https://developer.wordpress.org/reference/hooks/rest_this-post_type_query/) filters, which runs on `wp_template` and `wp_template_part` REST requests.

The reasoning behind running it every time is to check if a file has changed. E.g.:
https://github.com/WordPress/gutenberg/blob/759729cb22774acba6284045a3f2555c15fe16d4/lib/template-loader.php#L307-L310

I guess it makes sense, and it prevents us from, for example, creating the autodrafts once on theme activation.

Now, when we call this new `/page-templates` endpoint (btw: do we need one for template parts as well?), what happens with elements with a _file_, but without an _entity_ (aka without an autodraft)?
Are they going to be created as a side effect of `getEntityRecords`?
In other words, does `getEntityRecords` call a request on the post type, triggering the filter which creates the autodrafts needed?

Otherwise, we'd basically get a file name.
What do we do with it? 🤔

Reply:
> what happens with elements with a file, but without an entity (aka without an autodraft)?

Nothing, they are not created automagically, they won't be included in the `getEntityRecords` request. Template parts are created when requested via `getEntityRecords`, but not templates. The auto-draft is only created when we make a specifically request that template. That's why we need to expose an endpoint which returns all the file based templates.

Actually, I think it would be better to implement the same auto-drafting logic for templates. So we don't need two endpoints and it would save us a lot of hassle. I'll look into this.

Reply:
Replaced existing solution with magic. Templates are now automatically created as auto-drafts when requested. This simplified the code a lot!